### PR TITLE
fix(oracle): validate input with TO_TIMESTAMP_TZ and TO_DATE

### PR DESCRIPTION
**Pull request by WikiRik**
## Description of Changes

This adds input validation as a workaround to prevent SQL injection using `TO_TIMESTAMP` or `TO_DATE`. It's not pretty, but it should do the trick while we fix it properly in v7 where the internals should be more suitable. This fix is the least intrusive.

### CODE REVIEW

File: src/sql-string.js [hunk 1]
```diff
@@ -76,9 +77,54 @@ function escape(val, timeZone, dialect, format) {
       val = val.replace(/\0/g, '\\0');
     }
   } else if (dialect === 'oracle' && typeof val === 'string') {
-    if (val.startsWith('TO_TIMESTAMP') || val.startsWith('TO_DATE')) {
+    if (val.startsWith('TO_TIMESTAMP_TZ') || val.startsWith('TO_DATE')) {
+      const splitVal = val.split(/\(|\)/);
+    
+      if (splitVal.length !== 3 || splitVal[2] !== '') {
```
Comment (sdepold) [inline src/sql-string.js:83]: If I understand this right, we are checking here for the fact that we get something like `TO_DATE(col)` and have nothing coming after the closing paren? Please add a comment explaining this

File: src/sql-string.js [hunk 2]
```diff
@@ -76,9 +77,54 @@ function escape(val, timeZone, dialect, format) {
...
+      const insideParens = splitVal[1].trim();
+    
+      if (functionName !== 'TO_TIMESTAMP_TZ' && functionName !== 'TO_DATE') {
+        throw new Error('Invalid SQL function call. Expected TO_TIMESTAMP_TZ or TO_DATE.');
+      }
+    
+      const params = insideParens.split(',');
+    
+      if (params.length < 2) {
+        throw new Error('Invalid format for TO_TIMESTAMP_TZ or TO_DATE');
```
Comment (sdepold) [inline src/sql-string.js:97]: Does this mean that we only support the combination of fun + 2 params? Maybe change the error text to something like 

```
`Unexpected input received: ${val}\nSequelize supports TO_TIMESTAMP_TZ or TO_DATE exclusively with a combination of value and format.`
```

File: src/sql-string.js [hunk 3]
```diff
@@ -76,9 +77,54 @@ function escape(val, timeZone, dialect, format) {
...
+      }
+    
+      const dateValue = params[0].trim().replace(/'/g, '');
+      const formatValue = params[1].trim();
+    
+      if (functionName === 'TO_TIMESTAMP_TZ') {
+        const expectedFormat = "'YYYY-MM-DD HH24:MI:SS.FFTZH:TZM'";
+        if (formatValue !== expectedFormat) {
+          throw new Error(`Invalid format string for TO_TIMESTAMP_TZ. Expected format: ${ expectedFormat}`);
```
Comment (sdepold) [inline src/sql-string.js:106]: sooo. the second argument to the fun that is required as per the checks above always has to match `'YYYY-MM-DD HH24:MI:SS.FFTZH:TZM'`?

FYI: there is an extra space in before expectedFormat

File: src/sql-string.js [hunk 4]
```diff
@@ -76,9 +77,54 @@ function escape(val, timeZone, dialect, format) {
...
+        }
+    
+        const formattedDate = moment(dateValue).format('YYYY-MM-DD HH:mm:ss.SSS Z');
+        if (formattedDate !== dateValue) {
+          throw new Error("Invalid date value for TO_TIMESTAMP_TZ. Expected format: 'YYYY-MM-DD HH:mm:ss.SSS Z'");
+        }
+      } else if (functionName === 'TO_DATE') {
+        const expectedFormat = "'YYYY/MM/DD'";
+        if (formatValue !== expectedFormat) {
+          throw new Error(`Invalid format string for TO_DATE. Expected format: ${ expectedFormat}`);
```
Comment (sdepold) [inline src/sql-string.js:116]: extra space

File: src/sql-string.js (hunk 3 above)
Comment (WikiRik) [reply src/sql-string.js]: Yup, comment added and space removed

File: src/sql-string.js (hunk 2 above)
Comment (WikiRik) [reply src/sql-string.js]: Error text changed. Users should not receive this error text btw, unless they are doing some pre-processing or are trying to attack

File: src/sql-string.js (hunk 1 above)
Comment (WikiRik) [reply src/sql-string.js:86]: Yes, comment added

**Pull request merged**